# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent

### download and save locally
    with urllib.request.urlopen("https://data.bionet.nsw.gov.au/biosvcapp/odata/SpeciesNames",context=ssl.create_default_context(cafile=certifi.where())) as url:
        data = json.loads(url.read().decode())
    speciesnames = pd.json_normalize(data, record_path =['value'])
    speciesnames.to_csv(projectDir + 'source-data/nsw/nsw-species-names.csv', index=None)
    speciesnames

### 1. iNaturalist statuses

In [35]:
import pandas as pd
import sys
import os
projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
#projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
# taxastatus.head(3)
taxastatus

id taxon_id  user_id place_id source_id  \
0     166449    38493  1138587     7830             
1     234788   918383   702203     9994             
2     234789   918383   702203     7308             
3     166416  1033183  3669610     6825             
4     180721  1247288   222137     6825             
...      ...      ...      ...      ...       ...   
3448  165697  1182117  3669610    73684             
3449  130966   508985                               
3450  161226    50744   516268                      
3451  162783   924263   764897                      
3452  152695   779693   708886              16653   

                                              authority  \
0                    Flora and Fauna Guarantee Act 1988   
1                             Atlas of Living Australia   
2                             Atlas of Living Australia   
3                  NSW Office of Environment & Heritage   
4           NSW Threatened Species Scientific Committee   
...                                                 ...   
3448                              Australian Government   
3449                                                      
3450          DAWE Species Profile and Threats Database   
3451  Department of Biodiversity, Conservation and A...   
3452                                           QLD DEHP   

                     status  \
0                        CR   
1                       NT    
2                        LC   
3                        EN   
4                        vu   
...                     ...   
3448                     CR   
3449  critically endangered   
3450                     CR   
3451                     EN   
3452             endangered   

                                                    url description  \
0                                                                     
1     https://bie.ala.org.au/species/https://id.biod...               
2     https://bie.ala.org.au/species/https://id.biod...               
3     https://www.environment.nsw.gov.au/threateneds...               
4     https://www.environment.nsw.gov.au/topics/anim...               
...                                                 ...         ...   
3448  http://www.environment.gov.au/biodiversity/thr...               
3449  http://environment.gov.au/cgi-bin/sprat/public...               
3450  https://www.environment.gov.au/cgi-bin/sprat/p...               
3451  https://www.dpaw.wa.gov.au/images/documents/pl...               
3452  https://data.qld.gov.au/dataset/conservation-s...               

     geoprivacy  ...          genus specificEpithet infraspecificEpithet  \
0      obscured  ...      Eulamprus      kosciuskoi                        
1                ...   Chiloschista     phyllorhiza                        
2                ...   Chiloschista     phyllorhiza                        
3      obscured  ...       Eidothea      hardeniana                        
4      obscured  ...     Pomaderris         bodalla                        
...         ...  ...            ...             ...                  ...   
3448             ...    Achyranthes    margaretarum                        
3449   obscured  ...  Lichenostomus        melanops             cassidix   
3450   obscured  ...                                                       
3451   obscured  ...         Reedia       spathacea                        
3452   obscured  ...   Selenocosmia       crassipes                        

                  modified                   scientificName   taxonRank  \
0     2021-03-01T10:35:01Z             Eulamprus kosciuskoi     species   
1     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
2     2022-01-08T03:30:36Z         Chiloschista phyllorhiza     species   
3     2021-02-22T07:21:17Z              Eidothea hardeniana     species   
4     2021-08-27T06:18:35Z               Pomaderris bodalla     species   
...                    ...                              ...         ...   


In [36]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("NSW|New South Wales", ".nsw.")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
3       166416  1033183  3669610     6825             
2656    165059  1054498    58320     6825             
477     264941  1061113  3669610     6825             
371     180988  1070573  3669610     6825             
2514    160581  1076814   990532     6825             
...        ...      ...      ...      ...       ...   
1111    152247   906670   708886     6825     16650   
2637    164342   913024    58320     6825             
7       167825   953250   702203     6825             
836     155144   966856   708886     6825     16650   
2651    164615   973457   222137     6825             

                                              authority      status  \
3                  NSW Office of Environment & Heritage          EN   
2656               NSW Office of Environment & Heritage  Vulnerable   
477   New South Wales Biodiversity Conservation Act ...          EX   
371                NSW Office of Environment & Heritage          EN   
2514               NSW Office of Environment & Heritage  Vulnerable   
...                                                 ...         ...   
1111               NSW Office of Environment & Heritage  Endangered   
2637               NSW Office of Environment & Heritage  Vulnerable   
7     New South Wales Office of Environment and Heri...          VU   
836                NSW Office of Environment & Heritage  endangered   
2651        NSW Threatened Species Scientific Committee  Endangered   

                                                    url       description  \
3     https://www.environment.nsw.gov.au/threateneds...                     
2656  https://www.environment.nsw.gov.au/threatenedS...                     
477   https://bie.ala.org.au/species/https://id.biod...  Presumed Extinct   
371   https://www.environment.nsw.gov.au/threateneds...                     
2514  https://www.environment.nsw.gov.au/threateneds...                     
...                                                 ...               ...   
1111  https://www.environment.nsw.gov.au/resources/a...                     
2637  https://www.environment.nsw.gov.au/threatenedS...                     
7     https://www.environment.nsw.gov.au/threatenedS...        Vulnerable   
836   https://www.environment.nsw.gov.au/resources/a...                     
2651  https://www.environment.nsw.gov.au/topics/anim...                     

     geoprivacy  ...         genus specificEpithet infraspecificEpithet  \
3      obscured  ...      Eidothea      hardeniana                        
2656   obscured  ...  Prostanthera  cryptandroides       cryptandroides   
477        open  ...        Leuzea       australis                        
371              ...        Rotala      tripartita                        
2514   obscured  ...   Pterostylis       nigricans                        
...         ...  ...           ...             ...                  ...   
1111   obscured  ...     Caladenia        concolor                        
2637             ...     Symplocos     baeuerlenii                        
7                ...     Pultenaea          glabra                        
836    obscured  ...        Pteris   platyzomopsis                        
2651   obscured  ...       Calotis       pubescens                        

                  modified                              scientificName  \
3     2021-02-22T07:21:17Z                         Eidothea hardeniana   
2656  2020-12-17T03:53:45Z  Prostanthera cryptandroides cryptandroides   
477   2022-06-11T11:38:00Z                            Leuzea australis   
371   2021-09-21T05:48:55Z                           Rotala tripartita   
2514  2022-07-12T14:19:22Z                       Pterostylis nigricans   
...                    ...                                         ...   
1111  2022-06-10T07:36:31Z                          Caladenia concolor   
2637  2022-03-14T15:27:03Z                       Symplocos baeuerlenii   

### 2. iNaturalist taxonomy

In [37]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

Sensitive list: `geoprivacy` = `obscured`
overrides


In [38]:
%%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
#Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV

sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr487?max=10000&includeKVP=true")
sensitivelist = lf.kvp_to_columns(sensitivelist)
sensitivelist.to_csv(sourcedir + "nsw-ala-sensitive.csv", index=False)

conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr650?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist.to_csv(sourcedir + "nsw-ala-conservation.csv", index=False)

skipping # comment this line to download dataset from lists.ala.org.au the web and save locally


In [39]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "nsw-ala-sensitive.csv", dtype=str)
sensitivelist = sensitivelist.rename(columns={'T S Profile I D':'nsw_taxonID',
                                              'status':'nsw_status',
                                              'scientificName':'nsw_scientificName'})
sensitivelist

id                                  name  \
0    4040135              Callocephalon fimbriatum   
1    4040193                         Ninox strenua   
2    4040029                      Tyto tenebricosa   
3    4040138                       Ninox connivens   
4    4040203             Hoplocephalus bungaroides   
..       ...                                   ...   
192  4040083  Anthosachne kingiana subsp. kingiana   
193  4040198                       Caladenia tensa   
194  4040162                  Caladenia atroclavia   
195  4040056                  Prasophyllum sandrae   
196  4040139                Backhousia subargentea   

                        commonName                    nsw_scientificName  \
0               Gang-gang Cockatoo              Callocephalon fimbriatum   
1                     Powerful Owl           Ninox (Rhabdoglaux) strenua   
2                        Sooty Owl                      Tyto tenebricosa   
3                      Barking Owl        Ninox (Hieracoglaux) connivens   
4               Broad-headed Snake             Hoplocephalus bungaroides   
..                             ...                                   ...   
192     Phillip Island Wheat Grass  Anthosachne kingiana subsp. kingiana   
193  Erect Greencomb Spider Orchid                       Caladenia tensa   
194    Black-clubbed Spider-orchid                  Caladenia atroclavia   
195       Majors Creek Leek Orchid                  Prasophyllum sandrae   
196                 Giant Ironwood                Backhousia subargentea   

                                                  lsid dataResourceUid  \
0    https://biodiversity.org.au/afd/taxa/6c646af8-...           dr487   
1    https://biodiversity.org.au/afd/taxa/d1c5dee0-...           dr487   
2    https://biodiversity.org.au/afd/taxa/645b287c-...           dr487   
3    https://biodiversity.org.au/afd/taxa/bd332ca4-...           dr487   
4    https://biodiversity.org.au/afd/taxa/f579a6fc-...           dr487   
..                                                 ...             ...   
192  https://id.biodiversity.org.au/taxon/apni/5129...           dr487   
193  https://id.biodiversity.org.au/taxon/apni/5139...           dr487   
194  https://id.biodiversity.org.au/taxon/apni/5139...           dr487   
195                                      ALA_DR487_196           dr487   
196   https://id.biodiversity.org.au/node/apni/2888873           dr487   

                                             kvpValues   taxonRank   kingdom  \
0    [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
1    [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
2    [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
3    [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
4    [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species  Animalia   
..                                                 ...         ...       ...   
192  [{'key': 'taxonRank', 'value': 'Subspecies'}, ...  Subspecies   Plantae   
193  [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
194  [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
195  [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   
196  [{'key': 'taxonRank', 'value': 'Species'}, {'k...     Species   Plantae   

        class  ...           sourceStatus protectedInNSW sensitivityClass  \
0        Aves  ...             Vulnerable           True       Category 3   
1        Aves  ...             Vulnerable           True       Category 3   
2        Aves  ...             Vulnerable           True       Category 3   
3        Aves  ...             Vulnerable           True       Category 3   
4    Reptilia  ...             Endangered           True       Category 2   
..        ...  ...                    ...            ...              ...   
192     Flora  ...  Critically Endangered          False   

In [40]:
categorymapping = {
    "Category 3":"obscured",
    "Category 2":"obscured",
    "Category 1":"private"
}
sensitivelist['bionet_geoprivacy'] = sensitivelist['sensitivityClass'].str.strip().map(categorymapping).fillna('open') # map to dictionary
sensitivelist[['sensitivityClass','bionet_geoprivacy']].drop_duplicates()

sensitivityClass bionet_geoprivacy
0         Category 3          obscured
4         Category 2          obscured
143       Category 1           private

In [41]:
# Read conservation list data
conservationlist = pd.read_csv(sourcedir + "nsw-ala-conservation.csv", dtype=str)
conservationlist = conservationlist.rename(columns={'T S Profile I D':'nsw_taxonID',
                                              'status':'nsw_status',
                                              'scientificName':'nsw_scientificName'})
conservationlist['bionet_geoprivacy'] = conservationlist['sensitivityClass'].str.strip().map(categorymapping).fillna('open') # map to dictionary
conservationlist

id                           name             commonName  \
0     4028582                    Delma impar       Many-lined Delma   
1     4028454       Callocephalon fimbriatum     Gang-gang Cockatoo   
2     4028239            Cacophis harriettae    White-crowned Snake   
3     4028340        Litoria booroolongensis        Booroolong Frog   
4     4028512            Anthochaera phrygia      Regent Honeyeater   
...       ...                            ...                    ...   
1013  4028492    Muehlenbeckia sp. Mt Norman      Scrambling Lignum   
1014  4028995             Lobelia claviflora                    NaN   
1015  4029167    Fontainea sp. Coffs Harbour                    NaN   
1016  4029112           Melichrus gibberagee  Narrow-leaf Melichrus   
1017  4028270  Pittosporum sp. Coffs Harbour                    NaN   

                    nsw_scientificName  \
0                          Delma impar   
1             Callocephalon fimbriatum   
2                  Cacophis harriettae   
3              Litoria booroolongensis   
4     Anthochaera (Xanthomyza) phrygia   
...                                ...   
1013       Muehlenbeckia sp. Mt Norman   
1014                Lobelia claviflora   
1015       Fontainea sp. Coffs Harbour   
1016              Melichrus gibberagee   
1017     Pittosporum sp. Coffs Harbour   

                                                   lsid dataResourceUid  \
0     https://biodiversity.org.au/afd/taxa/51c700dd-...           dr650   
1     https://biodiversity.org.au/afd/taxa/6c646af8-...           dr650   
2     https://biodiversity.org.au/afd/taxa/ace6534f-...           dr650   
3     https://biodiversity.org.au/afd/taxa/6377e038-...           dr650   
4     https://biodiversity.org.au/afd/taxa/31869a0e-...           dr650   
...                                                 ...             ...   
1013                                     ALA_DR650_1014           dr650   
1014                                     ALA_DR650_1015           dr650   
1015                                     ALA_DR650_1016           dr650   
1016                                     ALA_DR650_1017           dr650   
1017                                     ALA_DR650_1018           dr650   

                                              kvpValues taxonRank   kingdom  \
0     [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
1     [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
2     [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
3     [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
4     [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species  Animalia   
...                                                 ...       ...       ...   
1013  [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
1014  [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
1015  [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
1016  [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   
1017  [{'key': 'taxonRank', 'value': 'Species'}, {'k...   Species   Plantae   

         class  ...           sourceStatus protectedInNSW sensitivityClass  \
0     Reptilia  ...             Vulnerable           True    Not Sensitive   
1         Aves  ...             Vulnerable           True       Category 3   
2     Reptilia  ...             Vulnerable           True    Not Sensitive   
3     Amphibia  ...             Endangered           True    Not Sensitive   
4         Aves  ...  Critically Endangered           True    Not Sensitive   
...        ...  ...                    ...            ...              ...   
1013     Flora  ...             Vulnerable          False    Not Sensitive   
1014     Flora  ...  Critically Endangered          False    Not Sensitive   
1015     Flora  ...  Critically Endangered          False    Not Sensitive   
1016     Flora  ...    

In [68]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['nsw_taxonID','name','nsw_status','lsid','bionet_geoprivacy','kingdom','class','family']],
                    conservationlist[['nsw_taxonID','name','nsw_status','lsid','bionet_geoprivacy','kingdom','class','family']]]).drop_duplicates()
statelist['nsw_taxonID'] = statelist['nsw_taxonID'].str[:-2]
statelist

nsw_taxonID                           name             nsw_status  \
0          10975       Callocephalon fimbriatum             Vulnerable   
1          10562                  Ninox strenua             Vulnerable   
2          10821               Tyto tenebricosa             Vulnerable   
3          10561                Ninox connivens             Vulnerable   
4          10413      Hoplocephalus bungaroides             Endangered   
...          ...                            ...                    ...   
1013       10547    Muehlenbeckia sp. Mt Norman             Vulnerable   
1014       20375             Lobelia claviflora  Critically Endangered   
1015       20376    Fontainea sp. Coffs Harbour  Critically Endangered   
1016       10521           Melichrus gibberagee             Endangered   
1017       20378  Pittosporum sp. Coffs Harbour  Critically Endangered   

                                                   lsid bionet_geoprivacy  \
0     https://biodiversity.org.au/afd/taxa/6c646af8-...          obscured   
1     https://biodiversity.org.au/afd/taxa/d1c5dee0-...          obscured   
2     https://biodiversity.org.au/afd/taxa/645b287c-...          obscured   
3     https://biodiversity.org.au/afd/taxa/bd332ca4-...          obscured   
4     https://biodiversity.org.au/afd/taxa/f579a6fc-...          obscured   
...                                                 ...               ...   
1013                                     ALA_DR650_1014              open   
1014                                     ALA_DR650_1015              open   
1015                                     ALA_DR650_1016              open   
1016                                     ALA_DR650_1017              open   
1017                                     ALA_DR650_1018              open   

       kingdom     class          family  
0     Animalia      Aves      Cacatuidae  
1     Animalia      Aves       Strigidae  
2     Animalia      Aves       Tytonidae  
3     Animalia      Aves       Strigidae  
4     Animalia  Reptilia        Elapidae  
...        ...       ...             ...  
1013   Plantae     Flora    Polygonaceae  
1014   Plantae     Flora   Campanulaceae  
1015   Plantae     Flora   Euphorbiaceae  
1016   Plantae     Flora       Ericaceae  
1017   Plantae     Flora  Pittosporaceae  

[1023 rows x 8 columns]

In [56]:
# %%script echo skipping # comment this line to download dataset
# retrieve binomial and trinomial names from GBIF
#statelist = statelist.rename(columns={'nsw_scientificName':'name'})
parsednames = lf.gbifparse(statelist)
parsednames.to_csv(sourcedir + "nsw-gbif.csv", index=False)

In [69]:
parsednames = pd.read_csv(sourcedir + "nsw-gbif.csv")
statelist = statelist.merge(parsednames[['scientificName','canonicalName','canonicalNameComplete','type','rankMarker']],how="left",left_on="name",right_on="scientificName")
numfullstatelist = len(statelist.index)
#conservationlist = conservationlist.rename(columns={'name':'nsw_scientificName'})
statelist['nsw_scientificName'] = statelist['canonicalName']
statelist

nsw_taxonID                           name             nsw_status  \
0          10975       Callocephalon fimbriatum             Vulnerable   
1          10562                  Ninox strenua             Vulnerable   
2          10821               Tyto tenebricosa             Vulnerable   
3          10561                Ninox connivens             Vulnerable   
4          10413      Hoplocephalus bungaroides             Endangered   
...          ...                            ...                    ...   
1018       10547    Muehlenbeckia sp. Mt Norman             Vulnerable   
1019       20375             Lobelia claviflora  Critically Endangered   
1020       20376    Fontainea sp. Coffs Harbour  Critically Endangered   
1021       10521           Melichrus gibberagee             Endangered   
1022       20378  Pittosporum sp. Coffs Harbour  Critically Endangered   

                                                   lsid bionet_geoprivacy  \
0     https://biodiversity.org.au/afd/taxa/6c646af8-...          obscured   
1     https://biodiversity.org.au/afd/taxa/d1c5dee0-...          obscured   
2     https://biodiversity.org.au/afd/taxa/645b287c-...          obscured   
3     https://biodiversity.org.au/afd/taxa/bd332ca4-...          obscured   
4     https://biodiversity.org.au/afd/taxa/f579a6fc-...          obscured   
...                                                 ...               ...   
1018                                     ALA_DR650_1014              open   
1019                                     ALA_DR650_1015              open   
1020                                     ALA_DR650_1016              open   
1021                                     ALA_DR650_1017              open   
1022                                     ALA_DR650_1018              open   

       kingdom     class          family                 scientificName  \
0     Animalia      Aves      Cacatuidae       Callocephalon fimbriatum   
1     Animalia      Aves       Strigidae                  Ninox strenua   
2     Animalia      Aves       Tytonidae               Tyto tenebricosa   
3     Animalia      Aves       Strigidae                Ninox connivens   
4     Animalia  Reptilia        Elapidae      Hoplocephalus bungaroides   
...        ...       ...             ...                            ...   
1018   Plantae     Flora    Polygonaceae    Muehlenbeckia sp. Mt Norman   
1019   Plantae     Flora   Campanulaceae             Lobelia claviflora   
1020   Plantae     Flora   Euphorbiaceae    Fontainea sp. Coffs Harbour   
1021   Plantae     Flora       Ericaceae           Melichrus gibberagee   
1022   Plantae     Flora  Pittosporaceae  Pittosporum sp. Coffs Harbour   

                  canonicalName      canonicalNameComplete        type  \
0      Callocephalon fimbriatum   Callocephalon fimbriatum  SCIENTIFIC   
1                 Ninox strenua              Ninox strenua  SCIENTIFIC   
2              Tyto tenebricosa           Tyto tenebricosa  SCIENTIFIC   
3               Ninox connivens            Ninox connivens  SCIENTIFIC   
4     Hoplocephalus bungaroides  Hoplocephalus bungaroides  SCIENTIFIC   
...                         ...                        ...         ...   
1018        Muehlenbeckia sp.Mt        Muehlenbeckia sp.Mt    INFORMAL   
1019         Lobelia claviflora         Lobelia claviflora  SCIENTIFIC   
1020         Fontainea sp.Coffs         Fontainea sp.Coffs    INFORMAL   
1021       Melichrus gibberagee       Melichrus gibberagee  SCIENTIFIC   
1022       Pittosporum sp.Coffs       Pittosporum sp.Coffs    INFORMAL   

     rankMarker         nsw_scientificName  
0           sp.   Callocephalon fimbriatum  
1           sp.              Ninox strenua  
2           sp.           Tyto tenebricosa  
3           sp.            Ninox connivens  
4           sp.  Hoplocephalus bungaroides  
...         ...                        ...  
1018        sp.        Muehlenbeckia sp.Mt  
1019        sp.         Lobelia claviflor

In [70]:
# Identify records that won't comply with iNaturalist species names
noncomply = statelist[statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID', 'BLACKLISTED']) ]
noncomply

nsw_taxonID                                      name  \
110        10706                Pterostylis sp. Botany Bay   
130        10718                      Pultenaea sp. Olinda   
132        10818                Typhonium sp. aff. brownii   
144        10242    Diuris sp. (Oaklands, D.L. Jones 5380)   
162        10717              Pultenaea sp. Genowlan Point   
171        20103                    Prasophyllum sp. Moama   
181        20266   Corunastylis sp. Charmhaven (NSW896673)   
829        10723                  Quassia sp. Moonee Creek   
920        10625  Phyllodes imperialis southern subspecies   
936        10317                     Eucalyptus sp. Cattai   
937        10900          Eucalyptus sp. Howes Swamp Creek   
944        20053              Astrotricha sp. Wallagaraugh   
956        20067   Bertya sp. (Chambigne NR, M. Fatemi 24)   
968        20147               Pelargonium sp. Striatellum   
971        20121    Bertya sp. (Clouds Creek, M. Fatemi 4)   
975        20140                   Hibbertia sp. Bankstown   
1004       20337                  Lenwebbia sp. Main Range   
1016       10028   Acacia terminalis subsp. Eastern Sydney   
1018       10547               Muehlenbeckia sp. Mt Norman   
1020       20376               Fontainea sp. Coffs Harbour   
1022       20378             Pittosporum sp. Coffs Harbour   

                 nsw_status  \
110              Endangered   
130              Endangered   
132              Endangered   
144              Endangered   
162   Critically Endangered   
171   Critically Endangered   
181   Critically Endangered   
829              Endangered   
920              Endangered   
936   Critically Endangered   
937              Endangered   
944              Endangered   
956              Endangered   
968              Endangered   
971              Endangered   
975   Critically Endangered   
1004  Critically Endangered   
1016             Endangered   
1018             Vulnerable   
1020  Critically Endangered   
1022  Critically Endangered   

                                                   lsid bionet_geoprivacy  \
110                                       ALA_DR487_111          obscured   
130                                       ALA_DR487_131          obscured   
132                                       ALA_DR487_133          obscured   
144                                       ALA_DR487_145          obscured   
162    https://id.biodiversity.org.au/node/apni/2891788          obscured   
171                                       ALA_DR487_172          obscured   
181   https://id.biodiversity.org.au/taxon/apni/5140...          obscured   
829    https://id.biodiversity.org.au/node/apni/2907911              open   
920                                       ALA_DR650_875              open   
936                                       ALA_DR650_899              open   
937                                       ALA_DR650_900              open   
944                                       ALA_DR650_908              open   
956                                       ALA_DR650_923              open   
968                                       ALA_DR650_941              open   
971                                       ALA_DR650_944              open   
975                                       ALA_DR650_950              open   
1004  https://id.biodiversity.org.au/taxon/apni/5129...              open   
1016                                     ALA_DR650_1011              open   
1018                                     ALA_DR650_1014              open   
1020                                     ALA_DR650_1016              open   
1022                                     ALA_DR650_1018              open   

       kingdom    class                  family  \
110    Plantae    Flora             Orchidaceae   
130    Plantae    Flora    Fabaceae (Faboideae)   
132    Plantae    Flora                 Araceae   
144    Plantae    Flora             Orchidaceae   
162    

In [71]:
# remove records that do not comply
statelist = statelist[~statelist['type'].isin(['INFORMAL','CULTIVAR','HYBRID', 'BLACKLISTED']) ]
statelist = pd.DataFrame(statelist[['nsw_taxonID','nsw_scientificName','nsw_status','lsid','bionet_geoprivacy']]).drop_duplicates()
statelist

nsw_taxonID               nsw_scientificName             nsw_status  \
0          10975         Callocephalon fimbriatum             Vulnerable   
1          10562                    Ninox strenua             Vulnerable   
2          10821                 Tyto tenebricosa             Vulnerable   
3          10561                  Ninox connivens             Vulnerable   
4          10413        Hoplocephalus bungaroides             Endangered   
...          ...                              ...                    ...   
1014       20369                Leionema westonii  Critically Endangered   
1015       10803  Thinornis cucullatus cucullatus  Critically Endangered   
1017       10111               Epacris gnidioides             Vulnerable   
1019       20375               Lobelia claviflora  Critically Endangered   
1021       10521             Melichrus gibberagee             Endangered   

                                                   lsid bionet_geoprivacy  
0     https://biodiversity.org.au/afd/taxa/6c646af8-...          obscured  
1     https://biodiversity.org.au/afd/taxa/d1c5dee0-...          obscured  
2     https://biodiversity.org.au/afd/taxa/645b287c-...          obscured  
3     https://biodiversity.org.au/afd/taxa/bd332ca4-...          obscured  
4     https://biodiversity.org.au/afd/taxa/f579a6fc-...          obscured  
...                                                 ...               ...  
1014                                     ALA_DR650_1008              open  
1015  https://biodiversity.org.au/afd/taxa/a1591e76-...              open  
1017  https://id.biodiversity.org.au/taxon/apni/5128...              open  
1019                                     ALA_DR650_1015              open  
1021                                     ALA_DR650_1017              open  

[1002 rows x 5 columns]

In [60]:
parsednames['type'].unique()

#cols for debugging

array(['SCIENTIFIC', 'INFORMAL', 'BLACKLISTED'], dtype=object)

In [72]:
# check for duplicates with conflicting information
dupinformation = statelist.groupby('nsw_taxonID').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)
dupinformation
#df.groupby('hash').filter(lambda group: len(group) > 1).sort('size', ascending=False)

Empty DataFrame
Columns: [nsw_taxonID, nsw_scientificName, nsw_status, lsid, bionet_geoprivacy]
Index: []

In [ ]:
# fix duplicate

### 4. Equivalent IUCN statuses

In [73]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['nsw_status'])['nsw_status'].count()

nsw_status
Critically Endangered    106
Endangered               414
Extinct                   72
Not Listed                 5
Vulnerable               405
Name: nsw_status, dtype: int64

In [74]:
iucnStatusMappings = {
    'critically endangered':'Critically Endangered',
    'endangered':'Endangered',
    'extinct':'Extinct',
    'not listed': 'Not Evaluated',
    'vulnerable': 'Vulnerable'
}


### 5. Determine best place ID to use

In [75]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6825

place_id  place_name       place_display_name 
6744      Australia        Australia                5
6825      New South Wales  New South Wales, AU    161
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [76]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['nsw_scientificName', 'nsw_taxonID','nsw_status','bionet_geoprivacy','lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='nsw_scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses

nsw_scientificName nsw_taxonID             nsw_status  \
138                      Acacia atrox       10003             Endangered   
617              Acacia baueri aspera       10005             Vulnerable   
97                 Acacia dangarensis       20028  Critically Endangered   
602                  Acacia pubifolia       10024             Endangered   
441                     Acacia ruppii       10027             Endangered   
...                               ...         ...                    ...   
1000                Leionema westonii       20369  Critically Endangered   
1001  Thinornis cucullatus cucullatus       10803  Critically Endangered   
1002               Epacris gnidioides       10111             Vulnerable   
1003               Lobelia claviflora       20375  Critically Endangered   
1004             Melichrus gibberagee       10521             Endangered   

     bionet_geoprivacy                                               lsid  \
138           obscured   https://id.biodiversity.org.au/node/apni/2894816   
617               open   https://id.biodiversity.org.au/node/apni/2892219   
97            obscured   https://id.biodiversity.org.au/node/apni/2915433   
602               open  https://id.biodiversity.org.au/taxon/apni/5129...   
441               open   https://id.biodiversity.org.au/node/apni/2901189   
...                ...                                                ...   
1000              open                                     ALA_DR650_1008   
1001              open  https://biodiversity.org.au/afd/taxa/a1591e76-...   
1002              open  https://id.biodiversity.org.au/taxon/apni/5128...   
1003              open                                     ALA_DR650_1015   
1004              open                                     ALA_DR650_1017   

     status_id        scientificName taxon_id user_id description iucn  \
138     152295          Acacia atrox   898643  708886               40   
617     159920  Acacia baueri aspera  1111698  702203               30   
97      152286    Acacia dangarensis   775137  708886               40   
602     162893      Acacia pubifolia   897779  702203               40   
441     162887         Acacia ruppii   578518  702203               40   
...        ...                   ...      ...     ...         ...  ...   
1000       NaN                   NaN      NaN     NaN         NaN  NaN   
1001       NaN                   NaN      NaN     NaN         NaN  NaN   
1002       NaN                   NaN      NaN     NaN         NaN  NaN   
1003       NaN                   NaN      NaN     NaN         NaN  NaN   
1004       NaN                   NaN      NaN     NaN         NaN  NaN   

                                 authority      status geoprivacy place_id  \
138   NSW Office of Environment & Heritage  endangered   obscured     6825   
617                        New South Wales          VU                6825   
97    NSW Office of Environment & Heritage  endangered   obscured     6825   
602                        New South Wales          EN                6825   
441                        New South Wales          EN                6825   
...                                    ...         ...        ...      ...   
1000                                   NaN         NaN        NaN      NaN   
1001                                   NaN         NaN        NaN      NaN   
1002                                   NaN         NaN        NaN      NaN   
1003                                   NaN         NaN        NaN      NaN   
1004                                   NaN         NaN        NaN      NaN   

       place_display_name  
138   New South Wales, AU  
617   New South Wales, AU  
97    New South Wales, AU  
602   New South Wales, AU  
441   New South Wales, AU  
...                   ...  
1000                  NaN  
1001                  NaN  
1002                  NaN  
1003                  NaN  
1004                  NaN  

[1005 rows x 16 co

In [77]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "New South Wales Office of Environment and Heritage"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://www.environment.nsw.gov.au/threatenedspeciesapp"
baseurl = "https://www.environment.nsw.gov.au/threatenedspeciesapp/profile.aspx?id="
biesearchurl = "https://bie.ala.org.au/species/" # eg + "https://id.biodiversity.org.au/node/apni/2894366"
mergedstatuses['new_url'] = mergedstatuses.apply(lambda x: biesearchurl + x['lsid'] if pd.isna(x['nsw_taxonID']) else baseurl + x['nsw_taxonID'],axis=1)
#mergedstatuses['new_url'] = baseurl + mergedstatuses['nsw_taxonID']

mergedstatuses.rename(columns={'bionet_geoprivacy':'new_geoprivacy'},inplace=True)
mergedstatuses['new_place_id'] = '6825'  # NEW SOUTH WALES
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['nsw_status'].fillna('Threatened')
mergedstatuses

nsw_scientificName nsw_taxonID             nsw_status  \
138                      Acacia atrox       10003             Endangered   
617              Acacia baueri aspera       10005             Vulnerable   
97                 Acacia dangarensis       20028  Critically Endangered   
602                  Acacia pubifolia       10024             Endangered   
441                     Acacia ruppii       10027             Endangered   
...                               ...         ...                    ...   
1000                Leionema westonii       20369  Critically Endangered   
1001  Thinornis cucullatus cucullatus       10803  Critically Endangered   
1002               Epacris gnidioides       10111             Vulnerable   
1003               Lobelia claviflora       20375  Critically Endangered   
1004             Melichrus gibberagee       10521             Endangered   

     new_geoprivacy                                               lsid  \
138        obscured   https://id.biodiversity.org.au/node/apni/2894816   
617            open   https://id.biodiversity.org.au/node/apni/2892219   
97         obscured   https://id.biodiversity.org.au/node/apni/2915433   
602            open  https://id.biodiversity.org.au/taxon/apni/5129...   
441            open   https://id.biodiversity.org.au/node/apni/2901189   
...             ...                                                ...   
1000           open                                     ALA_DR650_1008   
1001           open  https://biodiversity.org.au/afd/taxa/a1591e76-...   
1002           open  https://id.biodiversity.org.au/taxon/apni/5128...   
1003           open                                     ALA_DR650_1015   
1004           open                                     ALA_DR650_1017   

     status_id        scientificName taxon_id user_id description  ...  \
138     152295          Acacia atrox   898643  708886              ...   
617     159920  Acacia baueri aspera  1111698  702203              ...   
97      152286    Acacia dangarensis   775137  708886              ...   
602     162893      Acacia pubifolia   897779  702203              ...   
441     162887         Acacia ruppii   578518  702203              ...   
...        ...                   ...      ...     ...         ...  ...   
1000       NaN                   NaN      NaN     NaN         NaN  ...   
1001       NaN                   NaN      NaN     NaN         NaN  ...   
1002       NaN                   NaN      NaN     NaN         NaN  ...   
1003       NaN                   NaN      NaN     NaN         NaN  ...   
1004       NaN                   NaN      NaN     NaN         NaN  ...   

     geoprivacy place_id   place_display_name  \
138    obscured     6825  New South Wales, AU   
617                 6825  New South Wales, AU   
97     obscured     6825  New South Wales, AU   
602                 6825  New South Wales, AU   
441                 6825  New South Wales, AU   
...         ...      ...                  ...   
1000        NaN      NaN                  NaN   
1001        NaN      NaN                  NaN   
1002        NaN      NaN                  NaN   
1003        NaN      NaN                  NaN   
1004        NaN      NaN                  NaN   

                                          new_authority  \
138   New South Wales Office of Environment and Heri...   
617   New South Wales Office of Environment and Heri...   
97    New South Wales Office of Environment and Heri...   
602   New South Wales Office of Environment and Heri...   
441   New South Wales Office of Environment and Heri...   
...                                                 ...   
1000  New South Wales Office of Environment and Heri...   
1001  New South Wales Office of Environment and Heri...   
1002  New South Wales Office of Environment and Heri...   
1003  New South Wales Office of Environment and Heri...   
1004  New South Wales Office of Environment and Heri...   

                              

## Updates

In [78]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action             taxon_name      id taxon_id                 status  \
138  UPDATE           Acacia atrox  152295   898643             Endangered   
617  UPDATE   Acacia baueri aspera  159920  1111698             Vulnerable   
97   UPDATE     Acacia dangarensis  152286   775137  Critically Endangered   
602  UPDATE       Acacia pubifolia  162893   897779             Endangered   
441  UPDATE          Acacia ruppii  162887   578518             Endangered   
..      ...                    ...     ...      ...                    ...   
2    UPDATE       Tyto tenebricosa  152244    20422             Vulnerable   
111  UPDATE  Viola cleistogamoides  152321   566603             Endangered   
141  UPDATE       Wollemia nobilis  152312    49381  Critically Endangered   
698  UPDATE     Zieria involucrata  266172   604220             Endangered   
788  UPDATE       Zieria prostrata  165911   897289             Endangered   

    iucn_equivalent                                          authority  \
138      Endangered  New South Wales Office of Environment and Heri...   
617      Vulnerable  New South Wales Office of Environment and Heri...   
97       Endangered  New South Wales Office of Environment and Heri...   
602      Vulnerable  New South Wales Office of Environment and Heri...   
441      Vulnerable  New South Wales Office of Environment and Heri...   
..              ...                                                ...   
2        Vulnerable  New South Wales Office of Environment and Heri...   
111      Endangered  New South Wales Office of Environment and Heri...   
141      Endangered  New South Wales Office of Environment and Heri...   
698      Vulnerable  New South Wales Office of Environment and Heri...   
788      Vulnerable  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
138  https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
617  https://www.environment.nsw.gov.au/threateneds...       open     6825   
97   https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
602  https://www.environment.nsw.gov.au/threateneds...       open     6825   
441  https://www.environment.nsw.gov.au/threateneds...       open     6825   
..                                                 ...        ...      ...   
2    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
111  https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
141  https://www.environment.nsw.gov.au/threateneds...    private     6825   
698  https://www.environment.nsw.gov.au/threateneds...       open     6825   
788  https://www.environment.nsw.gov.au/threateneds...       open     6825   

      username                                        description  
138  peggydnew  Listed as Threatened - refer to https://www.en...  
617  peggydnew  Listed as Threatened - refer to https://www.en...  
97   peggydnew  Listed as Threatened - refer to https://www.en...  
602  peggydnew  Listed as Threatened - refer to https://www.en...  
441  peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
2    peggydnew  Listed as Threatened - refer to https://www.en...  
111  peggydnew  Listed as Threatened - refer to https://www.en...  
141  peggydnew  Listed as Threatened - refer to https://www.en...  
698  peggydnew  Listed as Threatened - refer to https://www.en...  
788  peggydnew  Listed as Threatened - refer to https://www.en...  

[149 rows x 12 columns]

## No status in iNaturalist via straight scientificName match
The NSW records that didn't match up to a status in iNaturalist

In [79]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="nsw_scientificName",right_on="scientificName")
noinatstatus

nsw_scientificName nsw_taxonID             nsw_status  \
0        Cyclopsitta diophthalma coxeni       10195  Critically Endangered   
1    Polytelis anthopeplus monarchoides       10644             Endangered   
2                    Genoplesium baueri       10875             Endangered   
3                      Lindsaea fraseri       10481             Endangered   
4             Callistemon linearifolius       10129             Vulnerable   
..                                  ...         ...                    ...   
851                   Leionema westonii       20369  Critically Endangered   
852     Thinornis cucullatus cucullatus       10803  Critically Endangered   
853                  Epacris gnidioides       10111             Vulnerable   
854                  Lobelia claviflora       20375  Critically Endangered   
855                Melichrus gibberagee       10521             Endangered   

    new_geoprivacy                                               lsid  \
0         obscured  https://biodiversity.org.au/afd/taxa/2c0e502e-...   
1         obscured  https://biodiversity.org.au/afd/taxa/1958a1ff-...   
2         obscured  https://id.biodiversity.org.au/taxon/apni/5140...   
3         obscured   https://id.biodiversity.org.au/node/apni/2906676   
4         obscured   https://id.biodiversity.org.au/node/apni/2897309   
..             ...                                                ...   
851           open                                     ALA_DR650_1008   
852           open  https://biodiversity.org.au/afd/taxa/a1591e76-...   
853           open  https://id.biodiversity.org.au/taxon/apni/5128...   
854           open                                     ALA_DR650_1015   
855           open                                     ALA_DR650_1017   

    status_id scientificName_x taxon_id user_id description  ...  \
0         NaN              NaN      NaN     NaN         NaN  ...   
1         NaN              NaN      NaN     NaN         NaN  ...   
2         NaN              NaN      NaN     NaN         NaN  ...   
3         NaN              NaN      NaN     NaN         NaN  ...   
4         NaN              NaN      NaN     NaN         NaN  ...   
..        ...              ...      ...     ...         ...  ...   
851       NaN              NaN      NaN     NaN         NaN  ...   
852       NaN              NaN      NaN     NaN         NaN  ...   
853       NaN              NaN      NaN     NaN         NaN  ...   
854       NaN              NaN      NaN     NaN         NaN  ...   
855       NaN              NaN      NaN     NaN         NaN  ...   

              class           order         family        genus  \
0              Aves  Psittaciformes  Psittaculidae  Cyclopsitta   
1              Aves  Psittaciformes  Psittaculidae    Polytelis   
2        Liliopsida     Asparagales    Orchidaceae  Genoplesium   
3    Polypodiopsida    Polypodiales   Lindsaeaceae     Lindsaea   
4               NaN             NaN            NaN          NaN   
..              ...             ...            ...          ...   
851             NaN             NaN            NaN          NaN   
852             NaN             NaN            NaN          NaN   
853   Magnoliopsida        Ericales      Ericaceae      Epacris   
854             NaN             NaN            NaN          NaN   
855             NaN             NaN            NaN          NaN   

    specificEpithet infraspecificEpithet              modified  \
0       diophthalma               coxeni  2021-11-24T00:57:49Z   
1       anthopeplus         monarchoides  2019-09-08T01:35:11Z   
2            baueri                  NaN  2020-02-18T23:46:33Z   
3           fraseri                  NaN  2020-11-16T20:28:38Z   
4               NaN                  NaN                   NaN   
..              ...                  ...                   ...   
851             NaN                  NaN                   NaN   
852             NaN                  NaN                 

In [80]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

nsw_scientificName nsw_taxonID             nsw_status  \
0        Cyclopsitta diophthalma coxeni       10195  Critically Endangered   
1    Polytelis anthopeplus monarchoides       10644             Endangered   
2                    Genoplesium baueri       10875             Endangered   
3                      Lindsaea fraseri       10481             Endangered   
5                Prostanthera marifolia       20101  Critically Endangered   
..                                  ...         ...                    ...   
846                  Homoranthus elusus       20362  Critically Endangered   
848                       Egernia roomi       20370  Critically Endangered   
849             Tympanocryptis osbornei       20380             Endangered   
850           Tympanocryptis mccartneyi       20379  Critically Endangered   
853                  Epacris gnidioides       10111             Vulnerable   

    new_geoprivacy                                               lsid  \
0         obscured  https://biodiversity.org.au/afd/taxa/2c0e502e-...   
1         obscured  https://biodiversity.org.au/afd/taxa/1958a1ff-...   
2         obscured  https://id.biodiversity.org.au/taxon/apni/5140...   
3         obscured   https://id.biodiversity.org.au/node/apni/2906676   
5         obscured   https://id.biodiversity.org.au/node/apni/2897722   
..             ...                                                ...   
846           open   https://id.biodiversity.org.au/node/apni/2898334   
848           open  https://biodiversity.org.au/afd/taxa/f6161e1f-...   
849           open  https://biodiversity.org.au/afd/taxa/42312895-...   
850           open  https://biodiversity.org.au/afd/taxa/88defa1e-...   
853           open  https://id.biodiversity.org.au/taxon/apni/5128...   

    status_id scientificName_x taxon_id user_id description  ...  \
0         NaN              NaN      NaN     NaN         NaN  ...   
1         NaN              NaN      NaN     NaN         NaN  ...   
2         NaN              NaN      NaN     NaN         NaN  ...   
3         NaN              NaN      NaN     NaN         NaN  ...   
5         NaN              NaN      NaN     NaN         NaN  ...   
..        ...              ...      ...     ...         ...  ...   
846       NaN              NaN      NaN     NaN         NaN  ...   
848       NaN              NaN      NaN     NaN         NaN  ...   
849       NaN              NaN      NaN     NaN         NaN  ...   
850       NaN              NaN      NaN     NaN         NaN  ...   
853       NaN              NaN      NaN     NaN         NaN  ...   

              class           order         family           genus  \
0              Aves  Psittaciformes  Psittaculidae     Cyclopsitta   
1              Aves  Psittaciformes  Psittaculidae       Polytelis   
2        Liliopsida     Asparagales    Orchidaceae     Genoplesium   
3    Polypodiopsida    Polypodiales   Lindsaeaceae        Lindsaea   
5     Magnoliopsida        Lamiales      Lamiaceae    Prostanthera   
..              ...             ...            ...             ...   
846   Magnoliopsida        Myrtales      Myrtaceae     Homoranthus   
848        Reptilia        Squamata      Scincidae         Egernia   
849        Reptilia        Squamata       Agamidae  Tympanocryptis   
850        Reptilia        Squamata       Agamidae  Tympanocryptis   
853   Magnoliopsida        Ericales      Ericaceae         Epacris   

    specificEpithet infraspecificEpithet              modified  \
0       diophthalma               coxeni  2021-11-24T00:57:49Z   
1       anthopeplus         monarchoides  2019-09-08T01:35:11Z   
2            baueri                  NaN  2020-02-18T23:46:33Z   
3           fraseri                  NaN  2020-11-16T20:28:38Z   
5         marifolia                  NaN  2020-02-19T17:43:25Z   
..              ...                  ...                   ...   
846          elusus                  NaN  2021-07-28T03:28:41Z   
848           roomi  

In [81]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['nsw_scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','nsw_scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'nsw_scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action                          taxon_name   id taxon_id  \
0      ADD      Cyclopsitta diophthalma coxeni  NaN   495917   
1      ADD  Polytelis anthopeplus monarchoides  NaN   728998   
2      ADD                  Genoplesium baueri  NaN   921754   
3      ADD                    Lindsaea fraseri  NaN  1081779   
5      ADD              Prostanthera marifolia  NaN   953191   
..     ...                                 ...  ...      ...   
846    ADD                  Homoranthus elusus  NaN  1244359   
848    ADD                       Egernia roomi  NaN  1031614   
849    ADD             Tympanocryptis osbornei  NaN   965994   
850    ADD           Tympanocryptis mccartneyi  NaN   965993   
853    ADD                  Epacris gnidioides  NaN   918333   

                    status iucn_equivalent  \
0    Critically Endangered      Vulnerable   
1               Endangered      Vulnerable   
2               Endangered      Vulnerable   
3               Endangered      Vulnerable   
5    Critically Endangered      Vulnerable   
..                     ...             ...   
846  Critically Endangered      Vulnerable   
848  Critically Endangered      Vulnerable   
849             Endangered      Vulnerable   
850  Critically Endangered      Vulnerable   
853             Vulnerable      Vulnerable   

                                             authority  \
0    New South Wales Office of Environment and Heri...   
1    New South Wales Office of Environment and Heri...   
2    New South Wales Office of Environment and Heri...   
3    New South Wales Office of Environment and Heri...   
5    New South Wales Office of Environment and Heri...   
..                                                 ...   
846  New South Wales Office of Environment and Heri...   
848  New South Wales Office of Environment and Heri...   
849  New South Wales Office of Environment and Heri...   
850  New South Wales Office of Environment and Heri...   
853  New South Wales Office of Environment and Heri...   

                                                   url geoprivacy place_id  \
0    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
1    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
2    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
3    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
5    https://www.environment.nsw.gov.au/threateneds...   obscured     6825   
..                                                 ...        ...      ...   
846  https://www.environment.nsw.gov.au/threateneds...       open     6825   
848  https://www.environment.nsw.gov.au/threateneds...       open     6825   
849  https://www.environment.nsw.gov.au/threateneds...       open     6825   
850  https://www.environment.nsw.gov.au/threateneds...       open     6825   
853  https://www.environment.nsw.gov.au/threateneds...       open     6825   

      username                                        description  
0    peggydnew  Listed as Threatened - refer to https://www.en...  
1    peggydnew  Listed as Threatened - refer to https://www.en...  
2    peggydnew  Listed as Threatened - refer to https://www.en...  
3    peggydnew  Listed as Threatened - refer to https://www.en...  
5    peggydnew  Listed as Threatened - refer to https://www.en...  
..         ...                                                ...  
846  peggydnew  Listed as Threatened - refer to https://www.en...  
848  peggydnew  Listed as Threatened - refer to https://www.en...  
849  peggydnew  Listed as Threatened - refer to https://www.en...  
850  peggydnew  Listed as Threatened - refer to https://www.en...  
853  peggydnew  Listed as Threatened - refer to https://www.en...  

[738 rows x 12 columns]

### Write updates and additions to file

In [82]:
pd.concat([updates,additions]).to_csv(sourcedir + "nsw.csv", index=False)

In [83]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

nsw_scientificName nsw_taxonID             nsw_status  \
4          Callistemon linearifolius       10129             Vulnerable   
13              Gentiana bredboensis       10346  Critically Endangered   
15        Gentiana wingecarribiensis       10347  Critically Endangered   
26           Myriophyllum implicatum       20146  Critically Endangered   
27      Lysimachia vulgaris davurica       10498             Endangered   
..                               ...         ...                    ...   
847                 Meridolum maryae       20363             Endangered   
851                Leionema westonii       20369  Critically Endangered   
852  Thinornis cucullatus cucullatus       10803  Critically Endangered   
854               Lobelia claviflora       20375  Critically Endangered   
855             Melichrus gibberagee       10521             Endangered   

    new_geoprivacy                                               lsid  \
4         obscured   https://id.biodiversity.org.au/node/apni/2897309   
13        obscured   https://id.biodiversity.org.au/node/apni/2900971   
15        obscured   https://id.biodiversity.org.au/node/apni/2915610   
26        obscured   https://id.biodiversity.org.au/node/apni/2889355   
27        obscured   https://id.biodiversity.org.au/node/apni/2901512   
..             ...                                                ...   
847           open  https://biodiversity.org.au/afd/taxa/02c8958c-...   
851           open                                     ALA_DR650_1008   
852           open  https://biodiversity.org.au/afd/taxa/a1591e76-...   
854           open                                     ALA_DR650_1015   
855           open                                     ALA_DR650_1017   

    status_id scientificName_x taxon_id user_id description  ... class order  \
4         NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
13        NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
15        NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
26        NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
27        NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
..        ...              ...      ...     ...         ...  ...   ...   ...   
847       NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
851       NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
852       NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
854       NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   
855       NaN              NaN      NaN     NaN         NaN  ...   NaN   NaN   

    family genus specificEpithet infraspecificEpithet modified  \
4      NaN   NaN             NaN                  NaN      NaN   
13     NaN   NaN             NaN                  NaN      NaN   
15     NaN   NaN             NaN                  NaN      NaN   
26     NaN   NaN             NaN                  NaN      NaN   
27     NaN   NaN             NaN                  NaN      NaN   
..     ...   ...             ...                  ...      ...   
847    NaN   NaN             NaN                  NaN      NaN   
851    NaN   NaN             NaN                  NaN      NaN   
852    NaN   NaN             NaN                  NaN      NaN   
854    NaN   NaN             NaN                  NaN      NaN   
855    NaN   NaN             NaN                  NaN      NaN   

    scientificName_y taxonRank references  
4                NaN       NaN        NaN  
13               NaN       NaN        NaN  
15               NaN       NaN        NaN  
26               NaN       NaN        NaN  
27               NaN       NaN        NaN  
..               ...       ...        ...  
847              NaN       NaN        NaN  
851              NaN       NaN        NaN  
852              NaN       NaN        NaN  
854              NaN       NaN        NaN  
85

### are there any that need to be removed?
nsw sensitive list count: 197
nsw inat statuses count: xx

updates to inat status: 152
additional inat status: 738
nsw statuses we can't find a taxon match for in iNaturalist: 139
total: 1395 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: xx-152=16 that may need checking

In [84]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
numnoncomply = len(noncomply.index)
numcomply = len(statelist.index)
numdupinfo = len(dupinformation.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Statelist merge': [numfullstatelist],
    'Species iNat Comply' : [numcomply],
    'Species iNat non-Comply': [numnoncomply],
    'Duplicate Information': [numdupinfo],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}
statsdf = pd.DataFrame(data=d)
statsdf


Sensitive  Conservation  Statelist merge  Species iNat Comply  \
0        197          1018             1023                 1002   

   Species iNat non-Comply  Duplicate Information  Updates  Additions  \
0                       21                      0      149        738   

   No Inat Status  Unknown to Inat  
0             856              118

In [ ]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]